### ISSUE TITLE
Eagerly provision connection pool

### ISSUE THREAD

**Issue by ziliangpeng** [brettwooldridge/HikariCP#1117]
#### Environment
```
HikariCP version: 2.7.1
JDK version     : 1.8.0_144
Database        : MySQL
Driver version  : all
```

I am building applications that serves high volume of requests in time sensitive manner. During bootstrapping/restart my applications, it is critical to make sure `HikariDataSource` to have fully provisioned at least >`minIdle` connections before I open the application to receive traffic.

It appears the HikariCP provision connections in a delay manner, relying on the `HouseKeeper` to provision connection asynchronously in background, and there's not a way to monitor whether at least `minidle` connections are ready.

It would be great to open an API in `HikariDataSource` to synchronously and eagerly provision connections so I can call this method during bootstrap and knowing my application will be ready for the adventure. I believe this would be a valuable functionality for high performance applications.

❤️🔦
-----------------------------------------------------------------------------------------
*Have you searched the CLOSED issues already? How about checking [stackoverflow](https://stackoverflow.com/search?q=HikariCP)?*

**Comment by brettwooldridge** [comment]
How many idle connections are configured?  It is true that only one connection is created synchronously during startup.  However, 100ms later the dedicated add-connection thread starts filling the pool up to ``minimumIdle`` level.

It has been my experience that in most applications pool initialization occurs very early in the startup sequence, and that by the time the startup is completed a few seconds later, the pool is already filled.  Anecdotally, my company’s server takes about 4-5sec to start, with the pool starting up between 1-2sec.  Filling the pool with 5 idle connections takes ~50ms, occurring 100ms after pool start and completing well before the end of server startup.

**Comment by ziliangpeng** [comment]
@brettwooldridge I've configured both the `maxPoolSize` and `minIdle` to 10.

Yes, by reading the code I learnt that the behavior is to do the first `fillPool` after 100ms, in the executorService (asynchronously).

I think the problem I have is that, in my Dropwizard app, once I've registered resources, my endpoints are open, and a very high volume of traffic comes in that requires db connections to do work. On an average case, only 1-2 connections are active, but if connections are not ready, these requests are blocked on `acquire`, resulting in quite a few timeouts.

What I'd like to do is to make sure all connections to be ready (properly created) before I mark the application as healthy, but I don't have a way to do it.

I think opening up the API as I suggested would let me do what I need. And it gives user more control on managing the lifecycle of the connection pool.

**Comment by brettwooldridge** [comment]
@ziliangpeng While I consider more how I want to officially support this capability, I have checked in code supporting an "undocumented" boolean system property, ``com.zaxxer.hikari.blockUntilFilled``, that works together with ``initializationFailTimeout``.

If the system property ``com.zaxxer.hikari.blockUntilFilled`` is set to ``"true"``, the pool will block during startup until ``minimumIdle`` connections exist in the pool **_or_** ``initializationFailTimeout`` milliseconds have passed.

**Comment by ziliangpeng** [comment]
Thank you! I'll try out the new feature.

**Comment by ziliangpeng** [comment]
@brettwooldridge which HikariCP version includes this change?

**Comment by brettwooldridge** [comment]
This version is not released yet. We typically wait until there are enough changes to justify a release.  I’ll review the changes since v2.7.8.

**Comment by ziliangpeng** [comment]
Is this change released? Which version would include this fix?

**Comment by brettwooldridge** [comment]
This is released, v3.2.0 is current.
